### PR TITLE
GH-1876: The projectstate should use relative URIs based on the location of the projectstate file

### LINKED ISSUE #1876: The projectstate should use relative URIs based on the location of the projectstate file
At the moment, the project state persists absolute file URIs which is a problem in case the project state would be published or in case a project folder is moved.

Also: Deal with the situation when a file mentioned in the project state does not exist anymore on disk.

### PR DESCRIPTION
**Pull request by mmews-n4**
closes #1876

### CODE REVIEW

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/ProjectStatePersister.java [hunk 1]
```diff
@@ -485,10 +497,32 @@ private File getDataFile(IProjectConfig project) {
...
+		return absoluteURI.toString();
+	}
+
+	private URI makeAbsolute(URI baseURI, String relativeString) {
+		if (relativeString.contains(":/")) {
+			return URI.createURI(relativeString);
+		}
+		URI relativeURI = URI.createFileURI(relativeString);
+		URI absoluteURI = relativeURI.resolve(baseURI);
+		return new FileURI(absoluteURI).toURI();
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/ProjectStatePersister.java:514]: Why converting it to a FileURI and back to an URI?

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/ide/xtext/server/ProjectStatePersisterTest.java [hunk 2]
```diff
@@ -47,6 +49,8 @@
 @SuppressWarnings({ "restriction", "deprecation" })
 public class ProjectStatePersisterTest {
 
+	static final URI BASE_URI = new FileURI(new File("/")).toURI();
```
Comment (mor-n4) [inline tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/ide/xtext/server/ProjectStatePersisterTest.java:52]: Wouldn't it be better to run the tests with a more realistic value?

**Review by mor-n4 — APPROVED** [review]
:+1:

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/ProjectStatePersister.java (hunk 1 above)
Comment (mmews-n4) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/ProjectStatePersister.java:514]: FileURI performs several assertions that I wanted to reuse here.

**Pull request merged**